internal/controller/clustersecurityintentbinding_controller.go

```diff
@@ -35,71 +39,184 @@ func (r *ClusterSecurityIntentBindingReconciler) Reconcile(ctx context.Context,
+	if _, ok := createEvent.Object.(*v1.ClusterNimbusPolicy); ok {
+		return false
 	}
+	return true
+}
+
+func (r *ClusterSecurityIntentBindingReconciler) updateFn(updateEvent event.UpdateEvent) bool {
+	// TODO: Handle update event for ClusterNimbusPolicy update so that reconciler don't process it
+	// twice.
+	return updateEvent.ObjectOld.GetGeneration() != updateEvent.ObjectNew.GetGeneration()
+}
+
+func (r *ClusterSecurityIntentBindingReconciler) deleteFn(deleteEvent event.DeleteEvent) bool {
+	obj := deleteEvent.Object
+	if _, ok := obj.(*v1.ClusterSecurityIntentBinding); ok {
+		return true
+	}
+	return ownerExists(r.Client, obj)
+}
 
-	existingClusterNp := &v1.ClusterNimbusPolicy{}
-	err = r.Get(ctx, types.NamespacedName{Name: req.Name}, existingClusterNp)
-	if err != nil && errors.IsNotFound(err) {
-		if err := r.Create(ctx, clusterNp); err != nil {
-			logger.Error(err, "failed to create ClusterNimbusPolicy", "ClusterNimbusPolicy.Name", clusterNp.Name)
-			return ctrl.Result{}, err
+func (r *ClusterSecurityIntentBindingReconciler) createOrUpdateCwnp(ctx context.Context, logger logr.Logger, req ctrl.Request) error {
+	// Always fetch the latest CRs so that we have the latest state of the CRs on the
+	// cluster.
+	var csib v1.ClusterSecurityIntentBinding
+	if err := r.Get(ctx, req.NamespacedName, &csib); err != nil {
+		logger.Error(err, "failed to fetch ClusterSecurityIntentBinding", "ClusterSecurityIntentBinding.Name", req.Name)
+		return err
+	}
+
+	var cwnp v1.ClusterNimbusPolicy
+	err := r.Get(ctx, req.NamespacedName, &cwnp)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return r.createCwnp(ctx, logger, csib)
 		}
-		return ctrl.Result{}, nil
+		logger.Error(err, "failed to fetch ClusterNimbusPolicy", "ClusterNimbusPolicy.Name", req.Name)
+	}
+	return r.updateCwnp(ctx, logger, csib)
+}
+
+func (r *ClusterSecurityIntentBindingReconciler) createCwnp(ctx context.Context, logger logr.Logger, csib v1.ClusterSecurityIntentBinding) error {
+	clusterNp := policybuilder.BuildClusterNimbusPolicy(ctx, logger, r.Client, r.Scheme, csib)
+	if clusterNp == nil {
+		return nil
 	}
-	clusterNp.ObjectMeta = existingClusterNp.ObjectMeta
+
+	if err := r.Create(ctx, clusterNp); err != nil {
+		logger.Error(err, "failed to create ClusterNimbusPolicy", "ClusterNimbusPolicy.Name", clusterNp.Name)
+		return err
+	}
+	logger.Info("ClusterNimbusPolicy created", "ClusterNimbusPolicy.Name", clusterNp.Name)
+
+	return nil
+}
+
+func (r *ClusterSecurityIntentBindingReconciler) updateCwnp(ctx context.Context, logger logr.Logger, csib v1.ClusterSecurityIntentBinding) error {
+	var existingCwnp v1.ClusterNimbusPolicy
+	if err := r.Get(ctx, types.NamespacedName{Name: csib.Name}, &existingCwnp); err != nil {
+		logger.Error(err, "failed to fetch ClusterNimbusPolicy", "ClusterNimbusPolicy.Name", csib.Name)
+		return err
+	}
+
+	clusterNp := policybuilder.BuildClusterNimbusPolicy(ctx, logger, r.Client, r.Scheme, csib)
+	if clusterNp == nil {
+		return nil
```

Comment:
Please return an error, and not nil

Reply:
I was not returning error due to CEL part. Anyways now I've addressed your comments.